File: deploy_kr260/tests/test_simple_axi_reg.c
/*
 * test_simple_axi_reg.c - Test del IP simple_axi_reg (P_02)
 *
 * Espera:
 *   0x00 reg0 RW
 *   0x04 reg1 RW
 *   0x08 reg2 RW
 *   0x0C reg3 RO = 0xBADC0DE5 (ID register)
 *
 * El test ID es el mas valioso: prueba que el PL esta programado con
 * NUESTRO bitstream (no basura ni PL anterior).
 *
 * Compilar: gcc -O2 -o test_simple_axi_reg test_simple_axi_reg.c
 * Ejecutar: sudo ./test_simple_axi_reg
 */

#include <stdio.h>
#include <stdint.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/mman.h>

#define BASE      0xA0000000UL  /* direccion asignada por Vivado */
#define SIZE      0x1000UL
#define ID_EXPECT 0xBADC0DE5

static volatile uint32_t *m;

static uint32_t rd(uint32_t o) { __sync_synchronize(); return m[o/4]; }
static void wr(uint32_t o, uint32_t v) { m[o/4] = v; __sync_synchronize(); }

int main(void) {
    int fd = open("/dev/mem", O_RDWR | O_SYNC);
    if (fd < 0) { perror("/dev/mem"); return 1; }

    void *p = mmap(NULL, SIZE, PROT_READ | PROT_WRITE, MAP_SHARED, fd, BASE);
    if (p == MAP_FAILED) { perror("mmap"); return 1; }
    m = p;

    /* Test 1: ID register */
    uint32_t id = rd(0x0C);
    int pass = 0, fail = 0;
    printf("reg3 (ID) = 0x%08X (esperado 0x%08X) %s\n",
           id, ID_EXPECT, id == ID_EXPECT ? "[OK]" : "[FAIL]");
    if (id == ID_EXPECT) pass++; else fail++;

    /* Test 2: reg0 RW */
    uint32_t vals[] = {0x11111111, 0xAAAAAAAA, 0x00000000, 0xFFFFFFFF};
    for (int r = 0; r < 3; r++) {
        for (size_t i = 0; i < sizeof(vals)/sizeof(vals[0]); i++) {
            wr(r*4, vals[i]);
            uint32_t got = rd(r*4);
            if (got == vals[i]) pass++;
            else {
                printf("reg%d write 0x%08X read 0x%08X [FAIL]\n", r, vals[i], got);
                fail++;
            }
        }
    }

    /* Test 3: reg3 RO - escribir NO debe cambiarlo */
    wr(0x0C, 0xDEADBEEF);
    uint32_t after = rd(0x0C);
    if (after == ID_EXPECT) {
        printf("reg3 RO intacto tras escritura %s\n", "[OK]");
        pass++;
    } else {
        printf("reg3 cambio a 0x%08X [FAIL] (deberia ser RO)\n", after);
        fail++;
    }

    printf("\n%d OK, %d FAIL\n", pass, fail);
    munmap(p, SIZE);
    close(fd);
    return fail == 0 ? 0 : 1;
}

File: deploy_kr260/tests/test_simple_axi_reg.py
#!/usr/bin/env python3
"""
test_simple_axi_reg.py - Test del IP simple_axi_reg (P_02)

Si has cargado el bitstream de P_02, el registro en offset 0x0C
siempre devuelve 0xBADC0DE5.

Uso: sudo python3 test_simple_axi_reg.py
"""

import mmap
import os
import struct
import sys

BASE = 0xA0000000
SIZE = 0x1000
ID_EXPECT = 0xBADC0DE5


def main():
    if os.geteuid() != 0:
        sys.exit("Requiere sudo")

    with open("/dev/mem", "r+b") as f:
        mem = mmap.mmap(f.fileno(), SIZE, offset=BASE)

    pass_n = fail_n = 0

    # Test ID
    mem.seek(0x0C)
    id_val = struct.unpack("<I", mem.read(4))[0]
    status = "OK" if id_val == ID_EXPECT else "FAIL"
    print(f"reg3 (ID) = 0x{id_val:08X} (esperado 0x{ID_EXPECT:08X}) [{status}]")
    (pass_n := pass_n + 1) if id_val == ID_EXPECT else (fail_n := fail_n + 1)

    # Test RW en reg0, reg1, reg2
    vals = [0x11111111, 0xAAAAAAAA, 0x00000000, 0xFFFFFFFF]
    for r in range(3):
        for v in vals:
            mem.seek(r * 4)
            mem.write(struct.pack("<I", v))
            mem.seek(r * 4)
            got = struct.unpack("<I", mem.read(4))[0]
            if got == v:
                pass_n += 1
            else:
                print(f"  reg{r} write 0x{v:08X} read 0x{got:08X} [FAIL]")
                fail_n += 1

    # Test reg3 es RO
    mem.seek(0x0C)
    mem.write(struct.pack("<I", 0xDEADBEEF))
    mem.seek(0x0C)
    after = struct.unpack("<I", mem.read(4))[0]
    if after == ID_EXPECT:
        print("reg3 RO intacto tras escritura [OK]")
        pass_n += 1
    else:
        print(f"reg3 cambio a 0x{after:08X} [FAIL]")
        fail_n += 1

    print(f"\n{pass_n} OK, {fail_n} FAIL")
    mem.close()
    sys.exit(0 if fail_n == 0 else 1)


if __name__ == "__main__":
    main()
